src/libstore/build/derivation-goal.cc

```diff
@@ -1452,12 +1483,28 @@ void DerivationGoal::waiteeDone(GoalPtr waitee, ExitCode result)
 {
     Goal::waiteeDone(waitee, result);
 
-    if (waitee->buildResult.success())
-        if (auto bfd = std::get_if<DerivedPath::Built>(&waitee->buildResult.path))
-            for (auto & [output, realisation] : waitee->buildResult.builtOutputs)
+    if (!useDerivation) return;
+    auto & fullDrv = *dynamic_cast<Derivation *>(drv.get());
+
+    auto * dg = dynamic_cast<DerivationGoal *>(&*waitee);
+    if (!dg) return;
+
+    auto outputs = fullDrv.inputDrvs.find(dg->drvPath);
+    if (outputs == fullDrv.inputDrvs.end()) return;
+
+    for (auto & outputName : outputs->second) {
+        auto buildResult = dg->getBuildResult(DerivedPath::Built {
+            .drvPath = dg->drvPath,
+            .outputs = OutputsSpec::Names { outputName },
+        });
+        if (buildResult.success()) {
+            for (auto & [output, realisation] : buildResult.builtOutputs) {
                 inputDrvOutputs.insert_or_assign(
-                    { bfd->drvPath, output.outputName },
+                    { dg->drvPath, output.outputName },
                     realisation.outPath);
+            }
+        }
+    }
 }
```

Comment:
This became more awkward after `buildResult` was encapsulated because one cannot easily index `builtOutputs` without knowing the hash modulo.

I think what this means is that
```
DrvPaths builtOutputs;
```
is the wrong field for `BuildResult`; instead of that `std::map<DrvOutput, Realisation>` it should be
```
std::map<std::string, Realisation> builtOutputs;
Hash drvHash;
```
since we are only returning realizations for a single top-level derivation.

Reply:
I pushed one more commit which addresses this awkwardness, getting rid of the double `for` loop.